Nodes reload request store password as SecureString

albertzaharovits opened the pull request:
This improvement was suggested by @jaymode in a [comment](https://github.com/elastic/elasticsearch/pull/30992#discussion_r192397683) in another review.

The node request, `NodesReloadSecureSettingsRequest`, broadcasting the password required to locally decrypt the keystore on each node, stores the password in a `SecureString` request field and not a plain `String` field.

Relates: https://github.com/elastic/elasticsearch/issues/29135
CC @elastic/es-security

elasticmachine commented:
Pinging @elastic/es-core-infra

albertzaharovits commented on server/src/main/java/org/elasticsearch/action/admin/cluster/node/reload/NodesReloadSecureSettingsRequest.java:
```diff
@@ -62,24 +68,60 @@ public ActionRequestValidationException validate() {
+        final byte[] bytes = Arrays.copyOfRange(byteBuffer.array(), byteBuffer.position(), byteBuffer.limit());
+        Arrays.fill(byteBuffer.array(), (byte) 0); // clear sensitive data
+        return bytes;
+    }
```
Copied from: https://github.com/elastic/elasticsearch/blob/113c1916eeffb8fb453d56c97942eefa19bcb087/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/CharArrays.java#L18

Alternatively, we could pull the `CharArrays` utils class out of `x-pack`. Yet this is the only use case in `server`. Moreover, `CharArrays` brings a methods that is narrowed to the security scope, namely `constantTimeEquals`.
Another alternative would be to have the `utf8BytesToChars` and `toUtf8Bytes` as static members of `SecureString` since right now uses are tied to `SecureString`.

All in all, I think copying them here for this sole use is the best alternative.

albertzaharovits commented on server/src/main/java/org/elasticsearch/action/admin/cluster/node/reload/NodesReloadSecureSettingsRequest.java:
```diff
@@ -62,24 +68,60 @@ public ActionRequestValidationException validate() {
+        final CharBuffer charBuffer = StandardCharsets.UTF_8.decode(byteBuffer);
+        final char[] chars = Arrays.copyOfRange(charBuffer.array(), charBuffer.position(), charBuffer.limit());
+        Arrays.fill(charBuffer.array(), (char) 0); // clear sensitive data
+        return chars;
```
Copied from: https://github.com/elastic/elasticsearch/blob/113c1916eeffb8fb453d56c97942eefa19bcb087/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/CharArrays.java#L18

See [previous comment](https://github.com/elastic/elasticsearch/pull/31261/files#r194657974) .

Some point worth raising here. In the original implementation (`CharArray#utf8BytesToChars`) the decoded `CharBuffer` is not cleared. It is `charBuffer.clear()` but should be `Arrays.fill(charBuffer.array(), (char) 0);`.
@jaymode Should I raise a PR to fix this?

jaymode replied:
I think it is worth raising a separate PR to fix this and move the methods as I said before. I also think that this method needs to handle some other cases such as my suggestion above about the hasArray() and other items.

jaymode commented on server/src/main/java/org/elasticsearch/action/admin/cluster/node/reload/NodesReloadSecureSettingsRequest.java:
```diff
@@ -62,24 +68,60 @@ public ActionRequestValidationException validate() {
+    private static byte[] toUtf8Bytes(char[] chars) {
+        final CharBuffer charBuffer = CharBuffer.wrap(chars);
+        final ByteBuffer byteBuffer = StandardCharsets.UTF_8.encode(charBuffer);
+        final byte[] bytes = Arrays.copyOfRange(byteBuffer.array(), byteBuffer.position(), byteBuffer.limit());
```
This code is dangerous so I think we should correct it. The char buffer is guaranteed to be a HeapCharBuffer by the javadocs; however there is no such guarantee about the ByteBuffer returned by `StandardCharsets.UTF_8.encode(charBuffer)`. We need to account for whether the bytebuffer has an array that can be accessed or not.

```java
final byte[] bytes;
if (byteBuffer.hasArray()) {
    // there is no guarantee that the byte buffers backing array is the right size so we need to make a copy
    bytes = Arrays.copyOfRange(byteBuffer.array(), byteBuffer.position(), byteBuffer.limit());
    Arrays.fill(byteBuffer.array(), (byte) 0); // clear sensitive data
} else {
    final int length = byteBuffer.limit() - byteBuffer.position();
    bytes = new byte[length];
    byteBuffer.get(bytes);
    // if the buffer is not read only we can reset and fill with 0's
    if (byteBuffer.isReadOnly() == false) {
         byteBuffer.clear(); // reset
         for (int i = 0; i < byteBuffer.limit(); i++) {
             byteBuffer.put((byte) 0);
         }
    }
}
```

albertzaharovits replied:
Thanks for the detailed snippet.

albertzaharovits commented:
Thank you for the feedback @jaymode .
I will open a follow-up PR to move the `NodesReloadSecureSettingsRequest#charsToUtf8Bytes` and `NodeesReloadSecureSettingsRequest#utf8BytesToChars` to `SecureString` after the feature branch is merged in.

There are two points worth bringing up after addressing the feedback, which I have pointed in the code.

albertzaharovits commented on server/src/main/java/org/elasticsearch/action/admin/cluster/node/reload/NodesReloadSecureSettingsRequestBuilder.java:
```diff
@@ -19,22 +19,66 @@
+        }
+        return this;
+    }
+
```
1/2: Since right now the keystore only uses the empty password, the previous implementation was sloppy and actually read the password as a REST parameter. This is terrible and has been corrected. It is now read from the request body.

jaymode replied:
We're going to need to figure out a way to filter this in the security auditing. In x-pack we introduced the concept of a `RestRequestFilter`. This concept might need to be promoted to core for this API.

albertzaharovits commented on server/src/main/java/org/elasticsearch/action/admin/cluster/node/reload/TransportNodesReloadSecureSettingsAction.java:
```diff
@@ -82,15 +83,15 @@ protected NodeRequest newNodeRequest(String nodeId, NodesReloadSecureSettingsReq
         final NodesReloadSecureSettingsRequest request = nodeReloadRequest.request;
         KeyStoreWrapper keystore = null;
-        try {
+        try (final SecureString secureSettingsPassword = request.secureSettingsPassword()) {
```
2/2 (1): The password is cleared on each `NodeRequest`.

albertzaharovits commented on server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestReloadSecureSettingsAction.java:
```diff
@@ -68,7 +72,8 @@ public RestResponse buildResponse(NodesReloadSecureSettingsResponse response, XC
                         builder.endObject();
-
+                        // clear password for the original request
+                        nodesRequest.secureSettingsPassword().close();
```
2/2 (2): The password is also cleared on the original `NodesRequest`.
The way everything fits together is as follows: The `NodeSRequest` has the password which has been read from the REST body. The password is stored as a `SecureString`. It is passed as a reference to all the constructed `NodeRequest`s. 
```
    public static class NodeRequest extends BaseNodeRequest {

        NodesReloadSecureSettingsRequest request;

        public NodeRequest() {
        }

        NodeRequest(String nodeId, NodesReloadSecureSettingsRequest request) {
            super(nodeId);
            this.request = request;
        }

        @Override
        public void readFrom(StreamInput in) throws IOException {
            super.readFrom(in);
            request = new NodesReloadSecureSettingsRequest();
            request.readFrom(in);
        }

        @Override
        public void writeTo(StreamOutput out) throws IOException {
            super.writeTo(out);
            request.writeTo(out);
        }
    }
```
ALL these requests are serialized (including the one for the local node, doing the broadcast) and their password is not cleared. The deserialized `NodeRequest`s will have their password cleared: https://github.com/elastic/elasticsearch/pull/31261/files#r195506031 .
The original password (read from the REST body) which was passed as a reference to all the broadcasted requests, which have been serialized, is cleared here.

All goes down in flames if the password in the `NodeRequest` for the local node is not copied (through serialize-deserialize) AND it is cleared before other requests are serialized.
Since ALL `NodeRequest`s are serialized this should not happen, right?

jaymode replied:
I believe everything you've said is correct

bellatrix1001 commented:
@jaymode How does one actually set this password via POST? Docs have no reference to secure_settings_password. Thanks!!

jaymode commented:
Password has been removed in a separate issue since we didn't have full support for it, so you cannot set it.
